### PR TITLE
HTTPoison.AsyncResponse id's can be {:maybe_redirect_response, status, headers, client}

### PR DESCRIPTION
**Pull request by vereis**
I ran into the following situation working with a `script.google.com/...` API:

POST-ing to a URL which returned a `302` meant that redirection was not followed because `:hackney` will only follow redirects if a POST request status was `304`, but if `follow_redirects: true` is given as a request option, `HTTPoison` returns: `{:ok, %HTTPoison.AsyncResponse{id: {:maybe_redirect, ..., ..., ...}}`

Passing `hackney: [force_redirects: true]` to the request to try to circumvent this does not work as expected, because `:hackney` doesn't follow the spec and change the `method` to `GET`, causing the request to return a `405`.

For my use case, handling the error is perfectly fine, but unfortunately dialyzer complains about the following match because the `HTTPoison.AsyncResponse`'s `id` is not a reference but instead just a tuple probably bubbling up from `:hackney`:

```elixir
    case HTTPoison.post(@url, params, [], follow_redirect: true) do
      {:ok, %HTTPoison.AsyncResponse{id: {:maybe_redirect, 302, headers, _}}} ->
        Logger.warning(...)
        :ok

      ... ->
    end
```

Not sure if this was known behaviour, but if so I've amended the type spec in this MR although perhaps one should return a different struct altogether?

### CODE REVIEW

**Comment by edgurgel** [comment]
@Vereis Thanks!

**Comment by Vereis** [comment]
Yeah that sounds better, I just didn't want to do too much out of nowhere, but yeah this seems more like a bug! Maybe I should have created an issue first 😅 

I think we can just handle where this gets returned and check if it looks like a `:maybe_redirect`, perhaps return something like `{:ok, %HTTPoison.MaybeRedirect{...}}` or an error? I'm not sure what you'd prefer in this case to be honest:

If we go with something like `{:ok, %HTTPoison.MaybeRedirect{}}` I guess we can return the location in this struct, but it'd be up to the user to manually try `GET`-ing this.

Perhaps since manual intervention is required this is moreso an `{:error, %HTTPoison.MaybeRedirect{}}` instead? What do you think?

If we wanted to handle this explicitly we could perhaps also add a function `HTTPoison.follow_redirect(%HTTPoison.MaybeRedirect{})` which would perform the `GET` for you, keeping the headers of the original request?

**Comment by edgurgel** [comment]
> If we go with something like {:ok, %HTTPoison.MaybeRedirect{}} I guess we can return the location in this struct, but it'd be up to the user to manually try GET-ing this.

I like this. Most people won't need to worry as it's not super easy to get to this case?

>If we wanted to handle this explicitly we could perhaps also add a function HTTPoison.follow_redirect(%HTTPoison.MaybeRedirect{}) which would perform the GET for you, keeping the headers of the original request?

Maybe we could wait for more feedback once we release the new struct?

What do you think?

**Comment by Vereis** [comment]
> I like this. Most people won't need to worry as it's not super easy to get to this case?

**Comment by Vereis** [comment]
> I like this. Most people won't need to worry as it's not super easy to get to this case?

Yeah exactly, this makes sense to me. I'll give it a shot and update this MR. Thanks for clarifying 💪

File: lib/httpoison/base.ex [hunk 1]
```diff
@@ -281,8 +302,21 @@ defmodule HTTPoison.Base do
...
       Issues an HTTP request using a `Request` struct.
 
-      This function returns `{:ok, response}` or `{:ok, async_response}` if the
-      request is successful, `{:error, reason}` otherwise.
+      This function returns `{:ok, response}`, `{:ok, async_response}`, or `{:ok, maybe_redirect}`
+      if the request is successful, `{:error, reason}` otherwise.
+
+      ## Redirect handling
+
+      If the option `:follow_redirect` is given, HTTP redirects are automatically follow if
```
Comment (Vereis) [inline lib/httpoison/base.ex:310]: Not sure if you'd rather have this information on this function or just add a note to see the documentation for the `MaybeRedirect` struct?

File: lib/httpoison/base.ex [hunk 2]
```diff
@@ -347,16 +382,29 @@ defmodule HTTPoison.Base do
...
       Options: see type `HTTPoison.Request`
 
-      This function returns `{:ok, response}` or `{:ok, async_response}` if the
-      request is successful, `{:error, reason}` otherwise.
+      This function returns `{:ok, response}`, `{:ok, async_response}`, or `{:ok, maybe_redirect}`
+      if the request is successful, `{:error, reason}` otherwise.
+
+      ## Redirect handling
+
+      If the option `:follow_redirect` is given, HTTP redirects are automatically follow if
```
Comment (Vereis) [inline lib/httpoison/base.ex:390]: Not sure if you'd rather have this information on this function or just add a note to see the documentation for the `MaybeRedirect` struct?

File: test/httpoison_base_test.exs [hunk 3]
```diff
@@ -569,6 +569,32 @@ defmodule HTTPoisonBaseTest do
              }
   end
 
+  test "request returns MaybeRedirect when passing follow_redirect option" do
```
Comment (Vereis) [inline test/httpoison_base_test.exs:572]: So I'm not sure what I'm doing wrong but I can't actually get `mix test` to run. Looking at the error it seems like `httparrot` requires some config to be set but there isn't any config for this app. 

How best would I go about testing this?

File: lib/httpoison/base.ex (hunk 2 above)
Comment (edgurgel) [reply lib/httpoison/base.ex:386]: I think it's fine like this 👍

File: test/httpoison_base_test.exs (hunk 3 above)
Comment (edgurgel) [reply test/httpoison_base_test.exs:572]: I will check it out!

Comment (edgurgel) [reply test/httpoison_base_test.exs:572]: Running the test the only errors I got were related to http versus https 🤔 If you change everything to `http://` it seems to work fine.

Comment (Vereis) [reply test/httpoison_base_test.exs:572]: Alright cool, I'll make that change then thanks 🙂

Comment (Vereis) [reply test/httpoison_base_test.exs:572]: I found the error; for the sake of asserting that the `redirect_url` does actually read from the response of the redirect headers I kept it as `https` vs `http` but fixed my mistake 😅

**Comment by edgurgel** [comment]
Nice one! I intend to release a new version soon. Cheers! 🎉

**Comment by Nitrino** [comment]
@edgurgel Can you please release a new version with these changes?

**Comment by edgurgel** [comment]
Yes! Absolutely! I completely forgot about this! 🤦  I will do it later today!

**Pull request merged**